common/dbconfig.cpp

```diff
@@ -0,0 +1,102 @@
+// Copyright 2021-present Pinterest. All Rights Reserved.
+//
+// @author Premkumar (prem@pinterest.com)
+//
+
+#include "common/dbconfig.h"
+#include "common/segment_utils.h"
+#include <utility>
+
+#include <gflags/gflags.h>
+#include <glog/logging.h>
+
+DECLARE_int32(replicator_replication_mode);
+
+namespace common {
+
+DBConfigManager* DBConfigManager::get() {
+  static DBConfigManager instance;
+  return &instance;
+}
+
+DBConfigManager::DBConfigManager() : dbConfig(new DBConfig()) {
+}
+
+bool DBConfigManager::loadJsonStr(const std::string& content) {
+  Json::Value root;
+  Json::Reader reader;
+  if (!reader.parse(content, root) || !root.isObject()) {
+    LOG(ERROR) << "Could not load config";
+    return false;
+  }
+
+  return loadJsonObject(root);
+}
+
+bool DBConfigManager::loadJsonObject(const Json::Value& root) {
+  try {
+    auto newConfig = parseConfig(root);
+    if (newConfig == nullptr) {
+      return false;
+    }
+
+    LOG(INFO) << "Successfully updated config";
+    std::atomic_exchange_explicit(&dbConfig, newConfig, std::memory_order_release);
+    fDataLoaded = true;
+    return true;
+  } catch (std::exception& ex) {
+    LOG(ERROR) << "Exception from parseConfig: " << ex.what();
+    return false;
+  }
+}
+
+std::shared_ptr<const DBConfig> DBConfigManager::parseConfig(const Json::Value& root) {
+  static const std::string ACK_MODE = "ack_mode";
+  static const std::string DATASET = "dataset";
+
+  DBConfig newDBConfig;
+
+  if (root.isMember(DATASET)) {
+    // parse dataset specific config
+    auto jsonDataSetConfigs = root[DATASET];
+
+    for (const auto& dataset : jsonDataSetConfigs.getMemberNames()) {
+      Config config;
+      const auto& jsonDataSetConfig = jsonDataSetConfigs[dataset];
+      bool isSet = false; /* creating this for future multiple assigns */
+
+      if(jsonDataSetConfig.isMember(ACK_MODE)) {
+        config.replication_mode = jsonDataSetConfig[ACK_MODE].asUInt();
+        isSet = true;
+      }
+
+      if (isSet) {
+        newDBConfig.dataSetConfigMap[dataset] = config;
+      }
+    }
+  }
+  return std::make_shared<const DBConfig>(std::move(newDBConfig));
+}
+
+/* conf getter functions */
+const Config* DBConfigManager::getConfig(const std::string& dbName) const {
```

Comment:
The returned Config pointer is not guaranteed to be valid, as it may outlive the dbConfig object.

Reply:
Have Changed this to Shared_ptr